horovod/run/gloo_run.py

```diff
@@ -0,0 +1,234 @@
+            rank += 1
+
+    if rank < np:
+        raise ValueError("Process number should not be larger than "
+                         "total available slots.")
+
+    # Fill in the local_size and cross_size because we can only know these number after
+    # allocation is done.
+    for alloc_item in alloc_list:
+        alloc_item.local_size = local_sizes[alloc_item.cross_rank]
+        alloc_item.cross_size = cross_sizes[alloc_item.local_rank]
+
+    return alloc_list
+
+
+def _launch_jobs(settings, host_alloc_plan, remote_host_names, _run_command):
+    """
+    executes the jobs defined by run command on hosts.
+    :param hosts_alloc: list of dict indicating the allocating info.
+    For example,
+        [{'Hostname':'worker-0', 'Rank': 0, 'Local_rank': 0, 'Cross_rank':0,
+            'Size':2, 'Local_size':1, 'Cross_size':2},
+        {'Hostname':'worker-1', 'Rank': 1, 'Local_rank': 0, 'Cross_rank':1,
+            'Size':2, 'Local_size':1, 'Cross_size':2}
+        ]
+    :type hosts_alloc: list(dict)
+    :param remote_host_names: names that are resolved to one of the addresses
+    of remote hosts interfaces.
+    :type remote_host_names: set
+    :param _run_command: command to execute
+    :type _run_command: string
+    :return:
+    :rtype:
+    """
+
+    def _exec_command(_command, _index, event_):
+        if settings.verbose:
+            print(_command)
+        try:
+            exit_code = safe_shell_exec.execute(_command, index=_index, event=event_)
+            if exit_code != 0:
+                print('Process {idx} exit with status code {ec}.'.format(idx=_index, ec=exit_code))
+        except Exception as e:
+            print('Exception happened during safe_shell_exec, exception '
+                  'message: {message}'.format(message=e))
+        return 0
+
+    if settings.ssh_port:
+        ssh_port_arg = "-p {ssh_port}".format(ssh_port=settings.ssh_port)
+    else:
+        ssh_port_arg = ""
+
+    # Create a event for communication between threads
+    event = threading.Event()
+
+    def set_event_on_sigterm(signum, frame):
+        event.set()
+
+    signal.signal(signal.SIGINT, set_event_on_sigterm)
+    signal.signal(signal.SIGTERM, set_event_on_sigterm)
+
+    args_list = []
+    for alloc_info in host_alloc_plan:
+        # generate env for rendezvous
+        horovod_rendez_env = 'HOROVOD_RANK={rank} HOROVOD_SIZE={size} ' \
+                             'HOROVOD_LOCAL_RANK={local_rank} HOROVOD_LOCAL_SIZE={local_size} ' \
+                             'HOROVOD_CROSS_RANK={cross_rank} HOROVOD_CROSS_SIZE={cross_size} ' \
+            .format(rank=alloc_info.rank, size=alloc_info.size,
+                    local_rank=alloc_info.local_rank, local_size=alloc_info.local_size,
+                    cross_rank=alloc_info.cross_rank, cross_size=alloc_info.cross_size)
+
+        host_name = alloc_info.hostname
+
+        env = os.environ.copy()
+        local_command = '{horovod_env} {env} {run_command}' .format(
```

Comment:
@tgaddair @zsh-thu, did we verify that this uses the current working directory of the user calling `horovodrun`?  E.g. if you call `horovodrun` from `/path/to/my/script`, would it `cd /path/to/my/script` before running the python code on other machines?  This is necessary to have parity with MPI behavior.

Reply:
Good question, I'll double check this tomorrow.

Reply:
No, that's not supported. Maybe we can get the absolute path from `$PWD` and add a `cd $PWD` command before running the actual command.

Reply:
#1298